test/gateway/eip20_gateway/progress_unstake.js

```diff
@@ -0,0 +1,535 @@
+        unlockSecret,
+      ),
+      'Invalid unlock secret.',
+    );
+
+  });
+
+  it('should fail when unstake message is undeclared', async function () {
+
+    await Utils.expectRevert(
+      gateway.progressUnstake(
+        unstakeMessage.messageHash,
+        unstakeMessage.unlockSecret,
+      ),
+      'Message on target status must be Declared.',
+    );
+
+  });
+
+  it('should fail when unstake message is already progressed', async function () {
+
+    await gateway.setInboxStatus(
+      unstakeMessage.messageHash,
+      MessageStatusEnum.Declared,
+    );
+
+    await gateway.progressUnstake(
+      unstakeMessage.messageHash,
+      unstakeMessage.unlockSecret,
+    );
+
+    await Utils.expectRevert(
+      gateway.progressUnstake(
+        unstakeMessage.messageHash,
+        unstakeMessage.unlockSecret,
+      ),
+      'Message on target status must be Declared.',
+    );
+
+  });
+
+  it('should fail for revoked redeem(unstake) message', async function () {
+
+    await gateway.setInboxStatus(
+      unstakeMessage.messageHash,
+      MessageStatusEnum.Revoked,
+    );
+
+    await Utils.expectRevert(
+      gateway.progressUnstake(
+        unstakeMessage.messageHash,
+        unstakeMessage.unlockSecret,
+      ),
+      'Message on target status must be Declared.',
+    );
+
+  });
+
+  it('should fail when redeem(unstake) message status is revocation declared',
+    async function () {
+
+    await gateway.setInboxStatus(
+      unstakeMessage.messageHash,
+      MessageStatusEnum.DeclaredRevocation,
+    );
+
+    await Utils.expectRevert(
+      gateway.progressUnstake(
+        unstakeMessage.messageHash,
+        unstakeMessage.unlockSecret,
+      ),
+      'Message on target status must be Declared.',
+    );
+
+  });
+
+  it('should fail when the reward amount is greater than the unstake amount',
+    async function () {
+
+    unstakeMessage.gasPrice = new BN(10000);
+    unstakeMessage.gasLimit = new BN(10000);
+
+    await setMessage();
+
+    await gateway.setInboxStatus(
+      unstakeMessage.messageHash,
+      MessageStatusEnum.Declared,
+    );
+
+    await Utils.expectRevert(
+      gateway.progressUnstake(
+        unstakeMessage.messageHash,
+        unstakeMessage.unlockSecret,
+      ),
+      'Reward amount must be less than redeem amount.',
+    );
+
+  });
+
+  it('should return correct "redeemAmount", "unstakeAmount" and ' +
+    '"rewardAmount" when gas price is zero', async function () {
+
+    unstakeMessage.gasPrice = new BN(0);
+
+    await setMessage();
+
+    await gateway.setInboxStatus(
+      unstakeMessage.messageHash,
+      MessageStatusEnum.Declared,
+    );
+
+    let result = await gateway.progressUnstake.call(
+      unstakeMessage.messageHash,
+      unstakeMessage.unlockSecret,
+    );
+
+    assert.strictEqual(
+      unstakeRequest.amount.eq(result.redeemAmount_),
+      true,
+      `Redeem amount ${result.redeemAmount_} must be equal to ${unstakeRequest.amount}`,
```

Comment:
`result.redeemAmount_` is a BN, it will not print readable value for reason. It should be converted to a number.  Please check at other places as-well.